C++ API parity: PoissonNLLLoss

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28783 C++ API parity: BCEWithLogitsLoss
* #28918 C++ API parity: isfinite
* #29000 C++ API parity: MarginRankingLoss
* **#28755 C++ API parity: PoissonNLLLoss**

Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

pbelevich committed:
C++ API parity: PoissonNLLLoss

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -134,6 +134,27 @@ inline Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets,
+    options.reduction() + " is not valid"
+  );
+  return torch::poisson_nll_loss(input, target, options.log_input(),
+    options.full(), options.eps(), get_enum(options.reduction()));
```
I think the value of `options.reduction()` directly maps to the enum values in `aten/src/ATen/core/Reduction.h`, and we should be able to remove the `get_enum` function.

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -1448,3 +1448,21 @@ TEST_F(FunctionalTest, CTCLoss) {
+    F::poisson_nll_loss(input, target,
+    PoissonNLLLossOptions().reduction(Reduction::Mean))));
+  ASSERT_THROWS_WITH(F::poisson_nll_loss(input, target,
+    PoissonNLLLossOptions().reduction(Reduction::END)), "not valid");
```
It would be awesome to add a "no-options" test case for `F::poisson_nll_loss` as well :D

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

pbelevich commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -216,6 +216,19 @@ inline Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets,
+    c10::get_if<enumtype::kNone>(&options.reduction()) ||
+    c10::get_if<enumtype::kMean>(&options.reduction()) ||
+    c10::get_if<enumtype::kSum>(&options.reduction()),
+    enumtype::get_enum_name(options.reduction()), " is not valid"
```
should we remove this check?

yf225 replied:
Yes I think we can remove it :D

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -2007,6 +2007,34 @@ TEST_F(ModulesTest, CTCLoss) {
 }
 
+TEST_F(ModulesTest, PoissonNLLLoss) {
+  using namespace at::Reduction;
```
nit: we can remove this line as well

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]

yf225 reviewed (approved):
@pbelevich Thanks so much for the awesome work!

pbelevich committed:
Update on "C++ API parity: PoissonNLLLoss"


Differential Revision: [D18202436](https://our.internmc.facebook.com/intern/diff/D18202436)

[ghstack-poisoned]
